### ISSUE TITLE
Support for zod superRefine in schema

### ISSUE THREAD

**Issue by superjose** [SeasonedSoftware/remix-forms#91]
Hi everyone 🤗. First of all, I'd like to thank you for developing this library. I just started using it, and it's pretty awesome!!! I love how fast you can come up with forms! Like seriously! Thanks 🤩!

I wanted to give a heads-up that I reached a small hiccup while working on dynamic forms. There's no support for `zod` objects that end with a `superRefine`! Since zod wraps them in a `ZodEffects` class its typing changes and it won't pass the TypeScript validation. 

I ended up digging for a couple of hours to see how we could fix this issue, and it shouldn't be that problematic (AFAIK, I didn't study remix-forms much). 

The solution seems to extract the schema from `ZodEffects,` like:
```js
const _schema = schema._def.typeName === 'ZodEffects' ? schema._def.schema : schema;
```

I created a small monkey patch directly to the index.js files which solved this issue. Note that there are typing issues that need to be solved, which I didn't check.

Affected Version:
remix-forms@0.18.2

```patch
diff --git a/node_modules/remix-forms/dist/index.js b/node_modules/remix-forms/dist/index.js
index aed6004..ce4e483 100644
--- a/node_modules/remix-forms/dist/index.js
+++ b/node_modules/remix-forms/dist/index.js
@@ -528,9 +528,10 @@ function Form({
     fieldErrorsComponent,
     Error2
   ]);
+  const _schema = schema._def.typeName === 'ZodEffects' ? schema._def.schema : schema;
   React5.useEffect(() => {
     var _a2;
-    for (const stringKey in schema.shape) {
+    for (const stringKey in _schema) {
       const key = stringKey;
       if (errors && ((_a2 = errors[key]) == null ? void 0 : _a2.length)) {
         try {
@@ -542,16 +543,19 @@ function Form({
   }, [errorsProp, unparsedActionData]);
   let autoFocused = false;
   let fields = [];
-  for (const stringKey in schema.shape) {
+  
+  for (const stringKey in _schema.shape) {
     const key = stringKey;
     const message = (_a = formErrors[key]) == null ? void 0 : _a.message;
-    const shape = schema.shape[stringKey];
+    const shape = _schema.shape[stringKey];
     const errorsArray = message && [message] || errors && errors[key];
     const fieldErrors = errorsArray && errorsArray.length ? errorsArray : void 0;
     const autoFocus = Boolean(fieldErrors && !autoFocused);
     if (autoFocus)
       autoFocused = true;
-    const { typeName, optional, nullable, getDefaultValue, enumValues } = shapeInfo(shape);
+      
+      const { typeName, optional, nullable, getDefaultValue, enumValues } = shapeInfo(shape);
+   
     const fieldType = typeName ? fieldTypes[typeName] : "string";
     const required = !(optional || nullable);
     const propOptions = options && options[key];
```
```patch
diff --git a/node_modules/remix-forms/dist/index.mjs b/node_modules/remix-forms/dist/index.mjs
index 4991000..bcfad61 100644
--- a/node_modules/remix-forms/dist/index.mjs
+++ b/node_modules/remix-forms/dist/index.mjs
@@ -470,6 +470,7 @@ function Form({
   const { formState } = form;
   const { errors: formErrors, isValid } = formState;
   const [disabled, setDisabled] = React5.useState(false);
+
   React5.useEffect(() => {
     const shouldDisable = mode === "onChange" || mode === "all" ? transition.state === "submitting" || !isValid : transition.state === "submitting";
     setDisabled(shouldDisable);
@@ -502,9 +503,11 @@ function Form({
     fieldErrorsComponent,
     Error
   ]);
+    const _schema = schema._def.typeName === 'ZodEffects' ? schema._def.schema : schema;
+
   React5.useEffect(() => {
     var _a2;
-    for (const stringKey in schema.shape) {
+    for (const stringKey in _schema) {
       const key = stringKey;
       if (errors && ((_a2 = errors[key]) == null ? void 0 : _a2.length)) {
         try {
@@ -516,10 +519,11 @@ function Form({
   }, [errorsProp, unparsedActionData]);
   let autoFocused = false;
   let fields = [];
-  for (const stringKey in schema.shape) {
+  
+  for (const stringKey in _schema.shape) {
     const key = stringKey;
     const message = (_a = formErrors[key]) == null ? void 0 : _a.message;
-    const shape = schema.shape[stringKey];
+    const shape = _schema.shape[stringKey];
     const errorsArray = message && [message] || errors && errors[key];
     const fieldErrors = errorsArray && errorsArray.length ? errorsArray : void 0;
     const autoFocus = Boolean(fieldErrors && !autoFocused);
```

### How to apply the patch
Note, for those lurking on this thread and who want to apply this in your package, please wait for the remix-forms team to find a fix! If by the time you're reading this and a patch hasn't happened, then you can:

```sh
yarn add patch-package
```

Create a `patches` folder at the root level and then add a `remix-forms+0.18.2.patch` file, as shown below
![image](https://user-images.githubusercontent.com/8230270/197292799-d8671a87-123f-4dcf-8ed0-027d1badc359.png)

Add in your `scripts` object in `package.json`:
```
scripts: {
 // other scripts,
 "postinstall": "patch-package"
}
```

And in your `Form` object, convert the `schema`  to unknown and cast it to a `typeof schema._def.schema` so you can have all the TypeScript goodness:

e.g:
```jsx

const schema =  z
    .object({
      planType: z.enum(["personal", "corporate"]),
      quantity: z.number().min(0),
    })
    .superRefine((arg, ctx) => {
      const isCorporate = arg.planType === "corporate";

      if (isCorporate && arg.quantity < 8) {
        ctx.addIssue({
          code: z.ZodIssueCode.too_small,
          minimum: env.corporateMin,
          message: `For corporate cards you must issue at least 8`,
          type: "number",
          inclusive: true,
          fatal: true,
          path: ["planType"],
        });
        return z.NEVER;
      }
    });

<Form
      schema={schema as unknown as typeof schema._def.schema}
    />
```

**Comment by danielweinmann** [comment]
@superjose, thanks for the kind words and for coming up with a solution for us!

@diogob used your solution as an inspiration for #92 and we just released `0.19.0` with it 🎉 we also added [examples to the site](https://remix-forms.seasoned.cc/examples/schemas/zod-effects). Can you test it and let us know if it worked for you?

Thanks @diogob for the PR and @superjose for the issue!

**Comment by superjose** [comment]
@danielweinmann and @diogob Leeeet's gooooooooooooooooooo! I'll be trying it later on tonight! I'll post back in case I find something!

Thank you so much 🤩
